compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -17,10 +20,87 @@ using llvm::APIntOps::GreatestCommonDivisor;
 
 namespace mlir::iree_compiler {
 
-std::optional<GPUMMASchedule>
+static int64_t calculateSharedMemoryUsedInBytes(const GPUMMASchedule &schedule,
+                                                int64_t lhsBitwidth,
+                                                int64_t rhsBitwidth) {
+  int64_t tileM = schedule.mSize * schedule.mTileCount * schedule.mWarpCount;
+  int64_t tileN = schedule.nSize * schedule.nTileCount * schedule.nWarpCount;
+  int64_t tileK = schedule.kSize * schedule.kTileCount;
+  return (tileM * tileK * lhsBitwidth + tileN * tileK * rhsBitwidth) / 8;
+}
+
+bool isValidSchedule(const GPUMatmulShapeType &problem,
+                     const GPUMMASchedule &schedule) {
+  bool isValidM = problem.mSize %
+                  (schedule.mSize * schedule.mTileCount * schedule.mWarpCount);
+  bool isValidN = problem.nSize %
+                  (schedule.nSize * schedule.nTileCount * schedule.nWarpCount);
+  bool isValidK = problem.kSize % (schedule.kSize * schedule.kTileCount) == 0;
+  return isValidN && isValidM && isValidK;
+}
+
+FailureOr<GPUMMASchedule> fitScheduleInSharedMemory(
+    const GPUMatmulShapeType &problem, ArrayRef<GPUMatmulShapeType> intrinsics,
+    GPUMMASchedule schedule, int64_t sharedMemLimitInBytes) {
+  int64_t lhsBitwidth =
+      intrinsics[schedule.index].aType.getIntOrFloatBitWidth();
+  int64_t rhsBitwidth =
+      intrinsics[schedule.index].bType.getIntOrFloatBitWidth();
+
+  while (!isValidSchedule(problem, schedule) ||
+         calculateSharedMemoryUsedInBytes(schedule, lhsBitwidth, rhsBitwidth) >
+             sharedMemLimitInBytes) {
+    LLVM_DEBUG({
+      llvm::dbgs() << "Shrinking schedule\n";
+      llvm::dbgs() << "mSize: " << schedule.mSize << "\n";
+      llvm::dbgs() << "nSize: " << schedule.nSize << "\n";
+      llvm::dbgs() << "kSize: " << schedule.kSize << "\n";
+      llvm::dbgs() << "mTileCount: " << schedule.mTileCount << "\n";
+      llvm::dbgs() << "nTileCount: " << schedule.nTileCount << "\n";
+      llvm::dbgs() << "kTileCount: " << schedule.kTileCount << "\n";
+      llvm::dbgs() << "mWarpCount: " << schedule.mWarpCount << "\n";
+      llvm::dbgs() << "nWarpCount: " << schedule.nWarpCount << "\n";
+    });
+
+    auto decrementIfPossible = [](int64_t &c) -> LogicalResult {
+      if (c <= 1) {
+        return failure();
+      }
+      --c;
+      return success();
+    };
+
+    // Attempt to shrink the schedule along one of the dimensions.
+    // TODO: A better solution should probably factor problem.mSize /
+    // (mWarpCount * mTileCount * mSize) and then pop off the smallest factors
+    // one at a time, preferably trying to keep the tile "generally square."
+    if (decrementIfPossible(schedule.mTileCount).succeeded()) {
```

Comment:
nit: AFAIK llvm prefers free functions for these
```suggestion
    if (succeeded(decrementIfPossible(schedule.mTileCount))) {
```